### PR TITLE
Update attachment link URL replacing logic

### PR DESCRIPTION
**Pull request by kidroca**
### Details

Update existing logic that changes URLs to resolve from API ROOT to handle more URL patterns.
URLs coming from these origins should be updated to resolve from API_ROOT
- staging
- prod
- absolute urls (URLs starting with a `/` that have no origin part)

### Fixed Issues

$ https://github.com/Expensify/App/issues/14227   
PROPOSAL: https://github.com/Expensify/App/issues/14227#issuecomment-1399323823

### Tests

- [ ] Verify staging and prod urls are resolved from API ROOT
    - Follow the steps from the issue description https://github.com/Expensify/App/issues/14227#issue-1529764776 in order to create attachment links from staging and prod. 
    - Inspect the Network tab or the HTML source (Chrome Dev tools -> Elements tab) to verify any attachment url (staging or prod) is resolved against the API ROOT of the current environment
- [ ] Verify absolute urls are resolved from API ROOT
    - At the moment the backend is not yet saving attachments using the absolute path, so we need to simulate the result
    - We need to inspect and modify Onyx data
    - Find an attachment action and modify the `html` of the message so that the `` part becomes ``
        <details>
        <summary>Example</summary>
        In Chrome dev console:
       
        ```js
        Onyx.connect({ key: 'reportActions_71955477', callback: actions => {
            const action = actions[932];
            action.message[0].html = action.message[0].html.replace('https://staging.expensify.com/', '/');
            Onyx.merge('reportActions_71955477', { 932: action });
        }});
        ```
        </details>
    - Now refresh the browser. The (modified) attachment should resolve and load successfully from API ROOT
- [ ] Verify all attachment usages are covered. All these places should resolve the attachment from API ROOT
    - attachment thumbnail in chat message comments 
    - attachment preview modal
    - attachment download button
    - use different attachment types - image formats / document formats

### Offline tests

Expected Offline behavior
- Images that were seen recently should be visible and render instantly (from cache)
- Images that haven't been already loaded would be blank

Expected slow internet behavior
- Images that were seen recently should be visible and render instantly (from cache)
- Other images would display a spinner / loader animation for a while until they load

### QA Steps

- [ ] Verify loading attachments still works
    - Follow the steps to from the issue description https://github.com/Expensify/App/issues/14227#issue-1529764776 in order to create attachment links from staging and prod. 
    - Inspect the Network tab or the HTML source (Chrome Dev tools -> Elements tab) to verify any attachment url (staging or prod) is resolved against the API ROOT of the current environment
    - Example: 
        1. Create attachment on PROD
        2. It would be saved with a `https://www.expensify.com/...` url
        3. Open staging web (and have Chrome dev tools open)
        4. See the same attachment 
        5. It should be resolved from `staging.expensify.com` 
- [ ] Upload different attachment types using staging and prod and verify they work
    - E.g. save a PDF, DOC, JPG, PNG on staging, save the same on prod. Verify they load successfully from both staging and prod
- [ ] Verify all attachment usages are covered. All these places should load and render successfully
    - attachment thumbnail in chat message comments 
    - attachment preview modal
    - attachment download button
    - use different attachment types - image formats / document formats

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

Screenshot 2023-01-23 at 11 03 48

Screenshot 2023-01-23 at 11 04 37
</details>

<details>
<summary>Mobile Web - Chrome</summary>

Screenshot 2023-01-24 at 0 50 01

</details>

<details>
<summary>Mobile Web - Safari</summary>

Screenshot 2023-01-23 at 11 06 45

</details>

<details>
<summary>Desktop</summary>

Screenshot 2023-01-23 at 11 07 56

</details>

<details>
<summary>iOS</summary>

Screenshot 2023-01-23 at 11 13 18

</details>

<details>
<summary>Android</summary>

Screenshot 2023-01-24 at 0 51 52

</details>

### CODE REVIEW

**Review by mountiny — COMMENTED** [review]
Code looks good to me in general! Great job! @mananjadhav will you be able to give this one a test?

**Comment by mananjadhav** [comment]
Hey hey!!! Not sure what went wrong here, I missed all the conversation on the issue 😥

@mountiny I'll review the conversation and also review the PR. Sorry for missing the proposal review.

**Review by mountiny — COMMENTED** [review]
Thanks @kidroca, waiting for the testing now

File: src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js [hunk 1]
```diff
@@ -30,20 +30,10 @@ const ImageRenderer = (props) => {
     //
     const isAttachment = Boolean(htmlAttribs[CONST.ATTACHMENT_SOURCE_ATTRIBUTE]);
     const originalFileName = htmlAttribs['data-name'];
-    let previewSource = htmlAttribs.src;
-    let source = isAttachment
+    const previewSource = replaceSourceOrigin(htmlAttribs.src);
+    const source = replaceSourceOrigin(isAttachment
```
[the remark this answers is not in the document]
Comment (kidroca) [reply src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js]: @mountiny 
> I think we could still keep a comment here noting what is being done here `replaceSourceOrigin` is not necessarily super clear

Renamed `replaceSourceOrigin` to `tryResolveUrlFromApiRoot`
Added more documentation about what it does and added comments on usages

The comments before usages don't explain **why**, but just that we **should** sometime change the URL
Any suggestions to improve them?

[the remark this answers is not in the document]
Comment (mountiny) [reply src/components/HTMLEngineProvider/HTMLRenderers/ImageRenderer.js]: I think its good for now, thanks!

**Review by mananjadhav — APPROVED** [review]

**Comment by mountiny** [comment]
@mananjadhav Will you be bale to do the reviewer checklist as well?

**Comment by mananjadhav** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/3069065/215811364-38d93620-6d09-4efd-94ae-2629ca996894.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/3069065/215811905-87c9f268-2089-4b7b-9f27-45d86c43b92d.mov

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/3069065/215812053-193a44c9-8428-46ef-8948-0323250a0bb9.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/3069065/215812155-4547b6d8-c175-42eb-8f9a-a2e289e8b103.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/3069065/215812228-18e142e9-9529-431d-b771-213e88921824.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/3069065/215812307-e3bf472f-bcf1-40ce-8fbb-f28dda1ab018.mov

</details>


-----

@mountiny Completed the review list, attachment testing took some time.

**Review by mountiny — APPROVED** [review]
Thank you @kidroca @mananjadhav

**Comment by mvtglobally** [comment]
We are facing issue with this one. Not sure if something we are doing is not right, but images  are opened from the wrong Environment
PDF which uploaded from staging not possible to open on the Prod 

https://user-images.githubusercontent.com/43995119/216183896-82867b63-2db5-43eb-81f2-7b0aa76b64ab.mov

**Comment by thienlnam** [comment]
Yes I'm running into the same issue - failure to open a PDF uploaded from staging in prod

Though other file types seem to work fine

**Comment by mountiny** [comment]
@mvtglobally @thienlnam Can we confirm this was not malfunctioning before already? I know there were some issues with these.

**Comment by thienlnam** [comment]
Not sure how we'd do that because now staging has the change and we can't check the previous version of staging.  Though I'm going to just create another issue for it https://github.com/Expensify/App/issues/14739 and move forward with deploy

**Comment by mvtglobally** [comment]
this PR supposed to fix it, no?

**Comment by mountiny** [comment]
@mvtglobally do you get anything in the console? Or any error with that PDF which is not able to load.

Yeah this PR should be loading the attachments from appropriate endpoints.

**Comment by kidroca** [comment]
@thienlnam @mountiny 

> Can we confirm this was not malfunctioning before already? I know there were some issues with these.

I had this issue from before the PR, I've posted about it here: https://github.com/Expensify/App/issues/14227#issuecomment-1384151477

I think at the moment only `staging` would behave correctly 

I should have described it in the testing steps, but since the fix is not yet on `prod` we can't expect that `prod` would behave correctly until the PR is deployed 
e.g. if you upload something on staging and view it from prod, it would still be loaded relative to staging, because the PR with the fix is not yet deployed on prod

**Comment by mountiny** [comment]
@kidroca Its deployed not https://github.com/Expensify/App/pull/14463#issuecomment-1412924589

@mvtglobally Would you be able to re-test with the latest App version, please? Thanks!

**Comment by mountiny** [comment]
Oh @kidroca tested [here](https://github.com/Expensify/App/issues/14227#issuecomment-1413820219) and it seems to work now! Great 🎉

**Pull request merged**